src/binary-reader.cc

```diff
@@ -963,28 +989,57 @@ Result BinaryReader::ReadFunctionBody(Offset end_offset) {
         CHECK_RESULT(ReadAlignment(&alignment_log2, "store alignment"));
         Address offset;
         CHECK_RESULT(ReadAddress(&offset, 0, "store offset"));
+        Index memidx;
+        if (alignment_log2 >> 6) {
+          ERROR_IF(!options_.features.multi_memory_enabled(),
+              "multi_memory not allowed");
+          CHECK_RESULT(ReadMemidx(&memidx, "store memidx"));
+          alignment_log2 = alignment_log2 & ((1 << 6) - 1);
+        } else {
+          memidx = 0;
+        }
 
-        CALLBACK(OnStoreExpr, opcode, alignment_log2, offset);
-        CALLBACK(OnOpcodeUint32Uint32, alignment_log2, offset);
+        CALLBACK(OnStoreExpr, opcode, alignment_log2, offset, memidx);
+        if (memidx == 0) {
+          CALLBACK(OnOpcodeUint32Uint32, alignment_log2, offset);
+        } else {
+          CALLBACK(OnOpcodeUint32Uint32Uint32, alignment_log2, offset, memidx);
+        }
         break;
       }
 
       case Opcode::MemorySize: {
-        uint8_t reserved;
-        CHECK_RESULT(ReadU8(&reserved, "memory.size reserved"));
-        ERROR_UNLESS(reserved == 0, "memory.size reserved value must be 0");
-        CALLBACK0(OnMemorySizeExpr);
-        CALLBACK(OnOpcodeUint32, reserved);
-        break;
+        if (!options_.features.multi_memory_enabled()) {
+          uint8_t reserved;
+          CHECK_RESULT(ReadU8(&reserved, "memory.size reserved"));
+          ERROR_UNLESS(reserved == 0, "memory.size reserved value must be 0");
+          CALLBACK(OnMemorySizeExpr, reserved);
+          CALLBACK(OnOpcodeUint32, reserved);
+          break; 
+        } else {
+          Index memidx;
+          CHECK_RESULT(ReadMemidx(&memidx, "memory.size memidx"));
+          CALLBACK(OnMemorySizeExpr, memidx);
+          CALLBACK(OnOpcodeUint32, memidx);
+          break;
+        }
       }
 
       case Opcode::MemoryGrow: {
-        uint8_t reserved;
-        CHECK_RESULT(ReadU8(&reserved, "memory.grow reserved"));
-        ERROR_UNLESS(reserved == 0, "memory.grow reserved value must be 0");
-        CALLBACK0(OnMemoryGrowExpr);
-        CALLBACK(OnOpcodeUint32, reserved);
-        break;
+        if (!options_.features.multi_memory_enabled()) {
+          uint8_t reserved;
+          CHECK_RESULT(ReadU8(&reserved, "memory.grow reserved"));
+          ERROR_UNLESS(reserved == 0, "memory.grow reserved value must be 0");
+          CALLBACK(OnMemoryGrowExpr, reserved);
+          CALLBACK(OnOpcodeUint32, reserved);
+          break; 
+        } else {
+          Index memidx;
+          CHECK_RESULT(ReadMemidx(&memidx, "memory.grow memidx"));
+          CALLBACK(OnMemoryGrowExpr, memidx);
+          CALLBACK(OnOpcodeUint32, memidx);
+          break;
```

Comment:
Can we above this if/else here and instead just do 

```
if (!options_.features.multi_memory_enabled())  {
  ERROR_UNLESS(reserved == 0, "memory.grow memory index must be 0");
}
```

Reply:
Same for MemoryInit and all the other memory operations below.  Hopefully we can avoid duplication that way?

Reply:
We kept these for similar reasons as OnOpcodeUint32Uint32Uint32. There are spec tests from WebAssembly/testsuite that require the module to be invalidated if the encoding of 0 memidx is anything other than 0x00. To remove this, we would need to require WebAssembly/testsuite to remove those tests.

Reply:
I see.

How about at least we do something like this to minimize duplication and make it clear precisely what the differences in the two code paths are:

```
Index memidx;
if (!options_.features.multi_memory_enabled())  {
  CHECK_RESULT(ReadU8(&reserved, "memory.grow reserved"));
  ERROR_UNLESS(memidx == 0, "memory.grow memory index must be 0");
} else {
  CHECK_RESULT(ReadMemidx(&memidx, "memory.grow memidx"));
}
CALLBACK(OnMemoryGrowExpr, memidx);
CALLBACK(OnOpcodeUint32, memidx);
```

Reply:
I think we can use this pattern here and elsewhere and it should reduce the size of this patch.